### DISCUSSION TITLE
How to handle errors

### DISCUSSION THREAD

**Discussion by pkrukp** [octokit/octokit.js#2039] [Q&A]
Hi,

I can't find any information about error handling.

For example I'm using `octokit.actions.downloadArtifact` - what happens when I can't download artifact, for example
* there's some network problem
* there's some authentication problem
* can't find artifact
* rate limit issue

Will it always throw `HttpError` or are there some cases when this (or any other) API returns some non-standard object?

Thanks

**Comment by gr2m** [comment]
You should always get an instance of [`RequestError`](https://github.com/octokit/request-error.js) in these cases. We are passing through whatever error we are getting from GitHub, which are documented at https://docs.github.com/en/rest/reference

authentication problems can be tricky. You might get a 404, not because the repository does not exist, but because the repository is private, and you don't have access with the current authentication. In same cases you get an explanatory explanation message, but not always. Sometimes the response headers include the required OAuth scopes for the endpoint along with the one the current token has (if it is an OAuth token), and sometimes it doesn't. GitHub's REST API is not as consistent as it could be

To avoid rate limits, I recommend using https://github.com/octokit/plugin-throttling.js/

**Reply by jakub-g** [reply]
Some documentation on error handling in README would be very useful because I didn't find any either.

For example when I do `let restResponse = await octo.rest.pulls.list(...)`, what happens in case of 500 error on backend? 

- will it reject the promise?
- or will it resolve, and I need to check for `if (restResponse.status !== 200)`?

Update: Found some info in `octokit/request` which says: [the promise will always be resolved](https://github.com/octokit/request.js#request). Maybe this info could be prominently cross-linked from this repo's README.

**Reply by MickeyPhoenix** [reply]
@jakub-g I'm not seeing anything in the page you linked, saying that "the promise will always be resolved." In fact, I'm seeing the opposite:

> If an error occurs, the promise is rejected with an error object containing 3 keys to help with debugging:
> 
> error.status The http response status code
> error.request The request options such as method, url and data
> error.response The http response object with url, headers, and data

And if you `await` a rejected `Promise`, it'll throw. So I believe you won't get the results you expect, by checking `if (restResponse.status !== 200)`.

**Reply by till** [reply]
Would anyone mind adding an example anyway? I don't want to ask for anything unreasonable, but am not writing JavaScript full-time and just wanted to try to build an action and this seems very complicated. E.g. I expected it to throw, but it doesn't seem to. Or I am using it wrong. Maybe others would find it useful as well?

The docs for Octokit seem really barebones still.

**Reply by gr2m** [reply]
The README of the `@octokit/request-error` module shows how the error object looks like:
https://github.com/octokit/request-error.js

An example of handling it would be

```js
try {
  // your code here that sends at least one Octokit request
  await octokit.request("GET /")
} catch (error) {
  // Octokit errors always have a `error.status` property which is the http response code
  if (error.status) {
    // handle Octokit error
    // see https://github.com/octokit/request-error.js
  } else {
    // handle all other errors
    throw error
  }
}
```

Does that make sense?

> I expected it to throw, but it doesn't seem to

Did you maybe forgot to use the `await` key word in your `try` block?

**Reply by till** [reply]
Maybe, I'll check tomorrow.

That example could go there, I can PR it later too.

**Reply by till** [reply]
I think I got confused as to when to `await` and when to `.then().catch()` etc.. But that seems to be working now.

**Reply by till** [reply]
I made a PR for the change: https://github.com/octokit/request-error.js/pull/334
